### PR TITLE
VPN-4986 Create a new Connection Manager Object with relevant states

### PR DESCRIPTION
**Pull request by Gela**
## Description

Create a new `ConnectionManager` class and move all functionalities from Controller to this new class. This component ConnectionManager sits between the frontend and the controller to handle all connectivity related logic and let the controller only handle activation and deactivation.
[This RFC](https://github.com/mozilla-mobile/mozilla-vpn-client/blob/main/docs/rfcs/0003-rfc-controller_refactor.md) goes into more details about the purpose and implementation details. 

In subsequent PRs, some of the functionality including activate() and deactivate() will be moved back to the Controller.

## Reference

    https://mozilla-hub.atlassian.net/browse/VPN-4986

## Checklist
    
- [ ] My code follows the style guidelines for this project
- [ ] I have not added any packages that contain high risk or unknown licenses (GPL,  LGPL, MPL, etc. consult with DevOps if in question)
- [ ] I have performed a self review of my own code
- [ ] I have commented my code PARTICULARLY in hard to understand areas
- [ ] I have added thorough tests where needed

### CODE REVIEW

File: src/apps/vpn/connectionmanager.h [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+  //
+  // At initialization time, these two member variables are set to
+  // MozillaVPN::serverData() to do not let not initialize.
+  //
+  // Please, do not use MozillaVPN::serverData() in the controller!
+  ServerData m_serverData;
+  ServerData m_nextServerData;
+
+  PingHelper m_ping_canary;
+  bool m_ping_received = false;
```
Comment (vinocher) [inline src/apps/vpn/connectionmanager.h:232]: Consider using lower camel casing for the names, for consistency with the C++ code base, now that it has moved to a new file.

File: src/apps/vpn/connectionmanager.h [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+#include "ipaddress.h"
+#include "loghandler.h"
+#include "models/server.h"
+#include "models/serverdata.h"
+#include "pinghelper.h"
+
+class ControllerImpl;
+class MozillaVPN;
+
+class ConnectionManager : public QObject, public LogSerializer {
```
Comment (vinocher) [inline src/apps/vpn/connectionmanager.h:23]: Consider adding a header comment describing purpose and usage of this class.

File: src/apps/vpn/connectionmanager.h
Comment (vinocher) [inline src/apps/vpn/connectionmanager.h:1]: The doc mentioned that this change would be behind a feature flag.

File: src/apps/vpn/captiveportal/captiveportaldetection.cpp
Comment (vinocher) [inline src/apps/vpn/captiveportal/captiveportaldetection.cpp:1]: Consider expanding the PR description to describe why this change is being done.

File: src/apps/vpn/connectionmanager.h [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+  //    ConnectionStateFirewall,
+  //    ConnectionStateCaptivePortal,
+  //    ConnectionStateCheckSubscription,
+  //    ConnectionStateUnstable,
+  //    ConnectionStateIdle,       // Used when there are no active probes
+  //    ongoing
+  //  };
+  //  Q_ENUM(ConnectionState)
+
+  // These are temporarily copied from the controller state to mirror them, but
```
Comment (vinocher) [inline src/apps/vpn/connectionmanager.h:40]: Do we use TODO flags along with tracking ticket numbers in the source code to track pending work? Or is this OK as is because we are sure that this change will be made soon?

File: src/apps/vpn/connectionmanager.cpp [hunk 4]
```diff
@@ -0,0 +1,1000 @@
...
+  Q_ASSERT(m_state == StateOn);
+  emit timeChanged();
+}
+
+void ConnectionManager::quit() {
+  logger.debug() << "Quitting";
+
+  if (m_state == StateInitializing || m_state == StateOff) {
+    emit readyToQuit();
+    return;
```
Comment (vinocher) [inline src/apps/vpn/connectionmanager.cpp:230]: Is it OK that m_nextStep is not set to Quit here? (This is old code, so just checking.)

File: src/apps/vpn/connectionmanager.h
Comment (Gela) [reply src/apps/vpn/connectionmanager.h:1]: I'd like to keep anything that changes functionality behind a feature flag until the work is complete but because this is simply moving the code from one class to another without changing anything, I don't think a feature flag is necessary.

File: src/apps/vpn/connectionmanager.h (hunk 3 above)
Comment (Gela) [reply src/apps/vpn/connectionmanager.h:40]: I'm not aware of a standard process for TODOs but I'd be happy to add a TODO flag for this. Let me know if the one I added looks okay or if there's a better way to do this

File: src/apps/vpn/connectionmanager.cpp (hunk 4 above)
Comment (Gela) [reply src/apps/vpn/connectionmanager.cpp:230]: m_nextStep is set to quit below this if statement. Are you asking if it should also be set to Quit within this?

Comment (vinocher) [reply src/apps/vpn/connectionmanager.cpp:230]: Yes, because the app's slot calls App::quit on the readyToQuit signal, and ConnectionManager's m_state is not updated to Quit because of the early return on line 230.

A change may not be necessary because the app is t, but updating the state may help to help future debugging during app shutdown. Feel free to ignore this comment.

**Comment by vinocher** [comment]
> I'd like to keep anything that changes functionality behind a feature flag until the work is complete but because this is simply moving the code from one class to another without changing anything, I don't think a feature flag is necessary. 

Yes, a feature flag for a large refactor like this is not easy and usually requires the feature check to be moved to a higher level, and for both the old and new code to be retained while the feature flag  remains in place. It's a trade-off between risk mitigation and source code clutter. If we feel that risk of this change is small because it mainly involves moving code, this seems OK. Thanks.

File: src/apps/vpn/connectionmanager.h
Comment (vinocher) [reply src/apps/vpn/connectionmanager.h:1]: Yes, a feature flag for a large refactor like this is not easy and usually requires the feature check to be moved to a igher, and for both the old and new code to be kept while the feature flag remains in place. It's a trade-off between risk mitigation and source code clutter. If we feel that risk of this change is small because it mainly involves moving code, this seems OK. Thanks.

File: src/apps/vpn/connectionmanager.h (hunk 3 above)
Comment (vinocher) [reply src/apps/vpn/connectionmanager.h]: The code base has some instances of the following pattern which references a ticket that tracks remaining work:
// TODO: Implement this once we update to Qt6.3 (VPN-3511)

File: src/apps/vpn/connectionmanager.cpp [hunk 5]
```diff
@@ -0,0 +1,999 @@
...
+
+  connect(&m_connectingTimer, &QTimer::timeout, this, [this]() {
+    m_enableDisconnectInConfirming = true;
+    emit enableDisconnectInConfirmingChanged();
+  });
+
+  connect(&m_handshakeTimer, &QTimer::timeout, this,
+          &ConnectionManager::handshakeTimeout);
+
+  LogHandler::instance()->registerLogSerializer(this);
```
Comment (mcleinman) [inline src/apps/vpn/connectionmanager.cpp:99]: do we also need to unregister the log serializer? This PR removes code doing this in controller.cpp.

File: src/apps/vpn/connectionmanager.cpp [hunk 6]
```diff
@@ -0,0 +1,999 @@
...
+  connect(LogHandler::instance(), &LogHandler::cleanupLogsNeeded, this,
+          &ConnectionManager::cleanupBackendLogs);
+
+  connect(this, &ConnectionManager::readyToServerUnavailable,
+          Tutorial::instance(), &Tutorial::stop);
+}
+
+void ConnectionManager::implInitialized(bool status, bool a_connected,
+                                        const QDateTime& connectionDate) {
+  logger.debug() << "Controller initialized with status:" << status
```
Comment (mcleinman) [inline src/apps/vpn/connectionmanager.cpp:183]:
```suggestion
  logger.debug() << "Connect manager initialized with status:" << status
```

?

File: src/apps/vpn/connectionmanager.cpp [hunk 7]
```diff
@@ -0,0 +1,999 @@
...
+  m_serverData.setExitServerPublicKey(
+      m_activationQueue.last().m_serverPublicKey);
+
+  m_pingReceived = false;
+  m_pingCanary.start(m_activationQueue.first().m_serverIpv4AddrIn, "0.0.0.0/0");
+  logger.info() << "Canary Ping Started";
+  activateNext();
+}
+
+void ConnectionManager::clearConnectedTime() {
```
Comment (mcleinman) [inline src/apps/vpn/connectionmanager.cpp:471]: Out of curiosity, why did you decide to reorder a bunch of the functions from how they were in `controller.cpp`? Totally fine, I'm just genuinely curious.

File: src/apps/vpn/connectionmanager.cpp [hunk 8]
```diff
@@ -0,0 +1,999 @@
...
+}
+
+void ConnectionManager::captivePortalPresent() {
+  if (m_portalDetected) {
+    return;
+  }
+  m_portalDetected = true;
+  logger.info() << "Captive-Portal Present, next activation will show prompt";
+}
+/////
```
Comment (mcleinman) [inline src/apps/vpn/connectionmanager.cpp:839]: Is this an artifact that should be removed?

**Review by mcleinman — COMMENTED** [review]
This is incredible, thank you! A few comments/questions.

**Review by vinocher — APPROVED** [review]

File: src/apps/vpn/connectionmanager.cpp (hunk 5 above)
Comment (Gela) [reply src/apps/vpn/connectionmanager.cpp:99]: yes! good catch

File: src/apps/vpn/connectionmanager.cpp (hunk 7 above)
Comment (Gela) [reply src/apps/vpn/connectionmanager.cpp:471]: I created the connection manager class and started moving functions over from the controller one by one and moved all dependencies for each function as needed so it ended up being a little different in ordering. If there's code that you think should be ordered differently let me know. In hindsight I could've just copied practically all of controller code over and worked backwards  🫠

File: src/apps/vpn/connectionmanager.cpp (hunk 4 above)
Comment (Gela) [reply src/apps/vpn/connectionmanager.cpp:234]: @vinocher Your logic makes a lot of sense to me so I went ahead and added your suggestion to the if statement above. If 'll double check with Naomi to make sure we're not missing something.

File: src/apps/vpn/connectionmanager.cpp (hunk 7 above)
Comment (mcleinman) [reply src/apps/vpn/connectionmanager.cpp]: Some of the ordering feels a bit less intuitive - activate/activateNext for example, but I'm certainly not going to block this on it. Thanks.

**Review by mcleinman — APPROVED** [review]
👍🏻

File: src/apps/vpn/connectionmanager.cpp (hunk 7 above)
Comment (Gela) [reply src/apps/vpn/connectionmanager.cpp]: If it helps at all, the next issue that is in progress is moving activate/deactivate and the corresponding helper back into the controller so it will be a little more intuitive soon :)
